### PR TITLE
ENH Restrict which repos get dispatch-ci

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/module-standardiser/issues/37

Note I ended up putting in some checking for is_meta_repo() aka .github repo which recently was added, this is required otherwise you cannot do a full dry-run

We want still want dispatch-ci on composer-plugins() specifically for vendor-plugin() which integrates with other things - https://github.com/silverstripe/vendor-plugin/actions/runs/7871071271

recipe-plugin() only does linting, so it's a bit pointless being on a cron like this though it's cleaner to just keep it on dispatch-ci - https://github.com/silverstripe/recipe-plugin/actions/runs/7854707347

should_use_dispatch_ci() will return as follows for the repos:

```
[0] => yes: silverstripe-maintenance
[1] => yes: silverstripe-composer-update-checker
[2] => yes: GridFieldBulkEditingTools
[3] => yes: cwp-agencyextensions
[4] => yes: cwp-starter-theme
[5] => no : developer-docs
[6] => yes: cwp-watea-theme
[7] => yes: silverstripe-elemental
[8] => yes: silverstripe-elemental-userforms
[9] => yes: silverstripe-admin
[10] => yes: silverstripe-asset-admin
[11] => yes: silverstripe-assets
[12] => yes: silverstripe-auditor
[13] => yes: silverstripe-blog
[14] => yes: silverstripe-campaign-admin
[15] => yes: silverstripe-cms
[16] => yes: silverstripe-config
[17] => yes: silverstripe-contentreview
[18] => yes: silverstripe-crontask
[19] => yes: silverstripe-documentconverter
[20] => yes: silverstripe-elemental-bannerblock
[21] => yes: silverstripe-elemental-fileblock
[22] => yes: silverstripe-environmentcheck
[23] => yes: silverstripe-errorpage
[24] => no : eslint-config
[25] => yes: silverstripe-externallinks
[26] => yes: silverstripe-framework
[27] => yes: silverstripe-graphql
[28] => yes: silverstripe-gridfieldqueuedexport
[29] => yes: silverstripe-hybridsessions
[30] => yes: silverstripe-iframe
[31] => yes: silverstripe-installer
[32] => yes: silverstripe-ldap
[33] => yes: silverstripe-linkfield
[34] => yes: silverstripe-lumberjack
[35] => yes: silverstripe-mimevalidator
[36] => yes: silverstripe-realme
[37] => yes: silverstripe-session-manager
[38] => yes: recipe-authoring-tools
[39] => yes: recipe-blog
[40] => yes: recipe-cms
[41] => yes: recipe-collaboration
[42] => yes: recipe-content-blocks
[43] => yes: recipe-core
[44] => yes: recipe-form-building
[45] => yes: recipe-plugin
[46] => yes: recipe-reporting-tools
[47] => yes: recipe-services
[48] => yes: silverstripe-registry
[49] => yes: silverstripe-reports
[50] => yes: silverstripe-restfulserver
[51] => yes: silverstripe-securityreport
[52] => yes: silverstripe-segment-field
[53] => yes: silverstripe-sharedraftcontent
[54] => yes: silverstripe-siteconfig
[55] => yes: silverstripe-sitewidecontent-report
[56] => yes: silverstripe-spamprotection
[57] => yes: silverstripe-staticpublishqueue
[58] => yes: silverstripe-subsites
[59] => yes: silverstripe-tagfield
[60] => yes: silverstripe-taxonomy
[61] => yes: silverstripe-textextraction
[62] => yes: silverstripe-userforms
[63] => yes: vendor-plugin
[64] => yes: silverstripe-versioned
[65] => yes: silverstripe-versioned-admin
[66] => yes: silverstripe-versionfeed
[67] => no : webpack-config
[68] => yes: silverstripe-simple
[69] => yes: silverstripe-advancedworkflow
[70] => yes: silverstripe-gridfieldextensions
[71] => yes: silverstripe-multivaluefield
[72] => yes: silverstripe-queuedjobs
[73] => yes: silverstripe-fluent
[74] => yes: silverstripe-mfa
[75] => yes: silverstripe-totp-authenticator
[76] => yes: silverstripe-webauthn-authenticator
[77] => yes: silverstripe-login-forms
[78] => yes: silverstripe-dynamodb
[79] => no : gha-ci
[80] => no : gha-run-tests
[81] => no : gha-generate-matrix
[82] => no : gha-auto-tag
[83] => no : gha-pull-request
[84] => no : gha-tag-release
[85] => no : gha-update-js
[86] => no : gha-keepalive
[87] => no : .github
[88] => no : gha-dispatch-ci
[89] => no : gha-issue
[90] => no : gha-merge-up
[91] => no : gha-action-ci
[92] => no : gha-gauge-release
[93] => no : gha-trigger-ci
[94] => no : markdown-php-codesniffer
[95] => no : silverstripe-standards
```

### CODE REVIEW

File: funcs_scripts.php [hunk 1]
```diff
@@ -216,6 +247,17 @@ function is_gha_repository()
...
+/**
+ * Determine if the module being processed should include a dispatch-ci.yml file
+ *
+ * Example usage:
+ * should_use_dispatch_ci()
+ */
+function should_use_dispatch_ci()
+{
+    return (is_module() || is_composer_plugin()) && !is_docs() && !is_gha_repository();
+}
```
Comment (GuySartorelli) [inline funcs_scripts.php:259]: Unless you can think of other times we'd use this, this logic should just be put directly into the `dispatch-ci.php` script. Otherwise every script would have it's own function in here.

File: update_command.php
Comment (GuySartorelli) [inline update_command.php:1]: What are the changes to this file for? Did you accidentally commit them? They're not mentioned in the PR description and don't seem related?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: update_command.php
Comment (emteknetnz) [reply update_command.php:1]: Ended up putting them in late after testing a full dry-run locally and seeing that the newly added .github repo failed

It's in description - `Note I ended up putting in some checking for is_meta_repo() aka .github repo which recently was added, this is required otherwise you cannot do a full dry-run`

.github repo has a `master` branch that doesn't match normal conventions

Comment (GuySartorelli) [reply update_command.php:1]: > Note I ended up putting in some checking for is_meta_repo()

I thought this was talking about the calls to that function in the scripts - thanks for clearing that up

File: funcs_scripts.php (hunk 1 above)
Comment (emteknetnz) [reply funcs_scripts.php]: Updated

File: scripts/cms-any/dispatch-ci.php [hunk 2]
```diff
@@ -32,6 +32,14 @@
         uses: silverstripe/gha-dispatch-ci@v1
 EOT;
 
-if (check_file_exists('.github/workflows/ci.yml')) {
-    write_file_even_if_exists('.github/workflows/dispatch-ci.yml', $content);
+$dispatchCiPath = '.github/workflows/dispatch-ci.yml';
+$ciPath = '.github/workflows/ci.yml';
+$shouldHaveDispatchCi = (is_module() || is_composer_plugin()) && !is_docs() && !is_gha_repository();
```
Comment (GuySartorelli) [inline scripts/cms-any/dispatch-ci.php:37]: Realistically I think `is_module() || is_composer_plugin()` is all that's needed here, but doesn't hurt to be explicit so it can stay as is.

File: update_command.php [hunk 3]
```diff
@@ -94,34 +94,41 @@
         } else {
             // get all branches
             $allBranches = explode("\n", cmd('git branch -r', $MODULE_DIR));
+            $allBranches = array_filter($allBranches, fn($branch) => !str_contains($branch, 'HEAD ->'));
```
Comment (GuySartorelli) [inline update_command.php:97]: What does this line do? It's not immediately clear and doesn't seem to be covered by the comment you made about the `.github` master branch.

File: update_command.php [hunk 4]
```diff
@@ -94,34 +94,41 @@
...
             $allBranches = array_map(fn($branch) => trim(str_replace('origin/', '', $branch)), $allBranches);
+            // reset index
+            $allBranches = array_values($allBranches);
```
Comment (GuySartorelli) [inline update_command.php:100]: Why? What does this actually do? What does "reset index" mean?

Comment (emteknetnz) [reply update_command.php:100]: It's because the array_filter() call above will remove some values, though it doesn't reset the int based index

array_values([1 => 'a', 4 => 'b']) becomes [0 = > 'a', 1 => 'b'];

File: update_command.php (hunk 3 above)
Comment (emteknetnz) [reply update_command.php:97]: It filters the array of string branch names and removes anything that contains the string `HEAD ->`

File: update_command.php (hunk 4 above)
Comment (GuySartorelli) [reply update_command.php:100]: Ahh, so "index" here is referring to the index of the array. Fair enough.

**Review by GuySartorelli — APPROVED** [review]
LGTM, I assume you've tested it.
